src/backend/renderer/mod.rs

```diff
@@ -0,0 +1,287 @@
+    fn height(&self) -> u32;
+}
+
+/// Abstraction of commonly used rendering operations for compositors.
+pub trait Renderer {
+    /// Error type returned by the rendering operations of this renderer.
+    type Error: Error;
+    /// Texture Handle type used by this renderer.
+    type TextureId: Texture;
+
+    /// Import a given bitmap into the renderer.
+    ///
+    /// Returns a texture_id, which can be used with `render_texture(_at)` or implementation-specific functions.
+    ///
+    /// If not otherwise defined by the implementation, this texture id is only valid for the renderer, that created it,
+    /// and needs to be freed by calling `destroy_texture` on this renderer to avoid a resource leak.
+    ///
+    /// This operation needs no bound or default rendering target.
+    #[cfg(feature = "image")]
+    fn import_bitmap<C: std::ops::Deref<Target = [u8]>>(
+        &mut self,
+        image: &image::ImageBuffer<image::Rgba<u8>, C>,
+    ) -> Result<Self::TextureId, Self::Error>;
+
+    /// Returns supported formats for shared memory buffers.
+    ///
+    /// Will always contain At least `Argb8888` and `Xrgb8888`.
+    #[cfg(feature = "wayland_frontend")]
+    fn shm_formats(&self) -> &[wl_shm::Format] {
+        // Mandatory
+        &[wl_shm::Format::Argb8888, wl_shm::Format::Xrgb8888]
+    }
+
+    /// Import a given buffer into the renderer.
+    ///
+    /// Returns a texture_id, which can be used with `render_texture(_at)` or implementation-specific functions.
+    ///
+    /// If not otherwise defined by the implementation, this texture id is only valid for the renderer, that created it.
+    ///
+    /// This operation needs no bound or default rendering target.
+    ///
+    /// The implementation defines, if the id keeps being valid, if the buffer is released,
+    /// to avoid relying on implementation details, keep the buffer alive, until you destroyed this texture again.
+    #[cfg(all(feature = "wayland_frontend", feature = "backend_egl"))]
+    fn import_buffer(
+        &mut self,
+        buffer: &wl_buffer::WlBuffer,
+        damage: Option<&Damage>,
+        egl: Option<&EGLBufferReader>,
+    ) -> Result<Self::TextureId, Self::Error>;
+
+    /// Initialize a rendering context on the current rendering target with given dimensions and transformation.
+    ///
+    /// This function *may* error, if:
+    /// - The given dimensions are unsuppored (too large) for this renderer
+    /// - The given Transformation is not supported by the renderer (`Transform::Normal` is always supported).
+    /// - There was a previous `begin`-call, which was not terminated by `finish`.
+    /// - This renderer implements `Bind`, no target was bound *and* has no default target.
+    /// - (Renderers not implementing `Bind` always have a default target.)
+    fn begin(
+        &mut self,
+        width: u32,
+        height: u32,
+        transform: Transform,
+    ) -> Result<(), <Self as Renderer>::Error>;
```

Comment:
General impression about this begin/end API. Would it make sense to instead build something with a closure, so that it cannot be misused?

Something like:

```rust
rendered.draw(|render_handle: &mut _| {
    // render_handle is some rendered-specific type that has the clear/render_at/etc... methods
});
```

Reply:
Might be a good idea.

Previously this trait featured an associated `Frame`-type, which was returned by `begin`, would `end` on drop and implemented the clear/render_at/etc functions. But that was an absolute lifetime nightmare.

I will see if the closure api works better and is flexible enough for anvil.

Reply:
That's the uncertainty I have with the closure yes, it's possible it'd not be flexible enough for the actual uses.